horovod/mxnet/mpi_ops.h

```diff
@@ -22,17 +22,53 @@
 #include <mxnet/engine.h>
 #include <mxnet/ndarray.h>
 
+#include "adapter.h"
+#include "tensor_util.h"
+
 namespace horovod {
 namespace mxnet {
 
 using namespace horovod::common;
 
-typedef ::mxnet::Engine Engine;
 typedef ::mxnet::NDArray NDArray;
-typedef ::mxnet::engine::CallbackOnComplete CallbackOnComplete;
-typedef ::mxnet::Engine::AsyncFn ExecFn;
+typedef ::mxnet::Engine::CallbackOnComplete CallbackOnComplete;
 typedef Request::RequestType OperationType;
-typedef std::shared_ptr<MXTemporaryBuffer<NDArray>> MXTempBufferShared;
+
+struct MpiOpsParam {
+  NDArray* input;
+  NDArray* output;
+  NDArray* cpu_tensor;
+  OperationType op_type;
+  std::string op_name;
+  int root_rank;
+
+  MpiOpsParam(NDArray* input, NDArray* output, NDArray* cpu_tensor,
```

Comment:
Add `const` to immutable arguments

Reply:
Actually adding `const` to these pointers requires cascading changes downstream where `const` does not apply (e.g. TensorUtil::Copy()). We have to keep it like this.